crepecake/src/main/java/com/google/cloud/tools/crepecake/cache/json/CacheMetadataLayerObjectTemplate.java

```diff
@@ -0,0 +1,164 @@
+
+  /**
+   * A list of image tags where the layer should exist at (i.e. the layer was pushed to the
+   * repository for that image).
+   */
+  private List<String> existsOn = new ArrayList<>();
+
+  /** Additional properties for the layer. */
+  @Nullable private PropertiesObject properties;
+
+  /**
+   * The reference for a layer consists of its size (in bytes), digest, and diff ID.
+   *
+   * @see Layer for details
+   */
+  private static class ReferenceObject extends JsonTemplate {
+
+    private long size;
+    private DescriptorDigest digest;
+    private DescriptorDigest diffId;
+  }
+
+  /** Extra properties for a layer. Only valid for application layers. */
+  private static class PropertiesObject extends JsonTemplate {
+
+    /** The paths to the source directories that the layer was constructed from. */
+    private List<String> sourceDirectories = new ArrayList<>();
+
+    /** The last time the layer was constructed. */
+    private long lastModifiedTime;
+  }
+
+  public CachedLayerType getType() {
+    return type;
+  }
+
+  public long getSize() {
+    return reference.size;
+  }
+
+  public DescriptorDigest getDigest() {
+    return reference.digest;
+  }
+
+  public DescriptorDigest getDiffId() {
+    return reference.diffId;
+  }
+
+  public List<String> getExistsOn() {
+    return existsOn;
+  }
+
+  public List<String> getSourceDirectories() throws CacheMetadataCorruptedException {
+    return getPropertiesNonnull().sourceDirectories;
+  }
+
+  public long getLastModifiedTime() throws CacheMetadataCorruptedException {
+    return getPropertiesNonnull().lastModifiedTime;
+  }
+
+  public CacheMetadataLayerObjectTemplate setType(CachedLayerType type) {
+    this.type = type;
+    return this;
+  }
+
+  public CacheMetadataLayerObjectTemplate setSize(long size) {
+    reference.size = size;
+    return this;
+  }
+
+  public CacheMetadataLayerObjectTemplate setDigest(DescriptorDigest digest) {
+    reference.digest = digest;
+    return this;
+  }
+
+  public CacheMetadataLayerObjectTemplate setDiffId(DescriptorDigest diffId) {
+    reference.diffId = diffId;
+    return this;
+  }
+
+  public CacheMetadataLayerObjectTemplate setExistsOn(List<String> existsOn) {
+    this.existsOn = existsOn;
+    return this;
+  }
+
+  public CacheMetadataLayerObjectTemplate setSourceDirectories(List<String> sourceDirectories) {
+    checkTypeValidForProperties();
+    createPropertiesIfNull().sourceDirectories = sourceDirectories;
+    return this;
+  }
+
+  public CacheMetadataLayerObjectTemplate setLastModifiedTime(long lastModifiedTime) {
+    checkTypeValidForProperties();
+    createPropertiesIfNull().lastModifiedTime = lastModifiedTime;
+    return this;
+  }
+
+  /** Returns properties if non-null; otherwise throws the appropriate exception. */
+  private PropertiesObject getPropertiesNonnull() throws CacheMetadataCorruptedException {
+    checkTypeValidForProperties();
+    if (null == properties) {
```

Comment:
no need to do this in Java since you can't use an assignment operator here
properties == null
is more conventional